java/ql/lib/semmle/code/java/frameworks/apache/Codec.qll

```diff
@@ -0,0 +1,135 @@
+
+private class CodecSummaryCsv extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "org.apache.commons.codec.binary;Base32;true;Base32;(int,byte[]);;Argument[1];Argument[-1];taint;",
+        "org.apache.commons.codec.binary;Base32;true;Base32;(int,byte[],boolean);;Argument[1];Argument[-1];taint;",
+        "org.apache.commons.codec.binary;Base32;true;Base32;(int,byte[],boolean,byte);;Argument[1];Argument[-1];taint;",
+        "org.apache.commons.codec.binary;Base32;true;Base32;(int,byte[],boolean,byte,CodecPolicy);;Argument[1];Argument[-1];taint;",
+        "org.apache.commons.codec.binary;Base64;true;Base64;(int,byte[]);;Argument[1];Argument[-1];taint;",
+        "org.apache.commons.codec.binary;Base64;true;Base64;(int,byte[],boolean);;Argument[1];Argument[-1];taint;",
+        "org.apache.commons.codec.binary;Base64;true;Base64;(int,byte[],boolean,CodecPolicy);;Argument[1];Argument[-1];taint;",
+        "org.apache.commons.codec.binary;Base64;true;decodeBase64;(String);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;Base64;true;decodeBase64;(byte[]);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;Base64;true;encodeBase64;(byte[]);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;Base64;true;encodeBase64;(byte[],boolean);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;Base64;true;encodeBase64;(byte[],boolean,boolean);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;Base64;true;encodeBase64;(byte[],boolean,boolean,int);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;Base64;true;encodeBase64Chunked;(byte[]);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;Base64;true;encodeBase64String;(byte[]);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;Base64;true;encodeBase64URLSafe;(byte[]);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;Base64;true;encodeBase64URLSafeString;(byte[]);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;BaseNCodec;true;decode;(String);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;BaseNCodec;true;encode;(byte[],int,int);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;BaseNCodec;true;encodeAsString;(byte[]);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;BaseNCodec;true;encodeToString;(byte[]);;Argument[0];ReturnValue;taint;",
+        "org.apache.commons.codec.binary;BaseNCodecOutputStream;true;BaseNCodecOutputStream;(OutputStream,BaseNCodec,boolean);;Argument[1];Argument[-1];taint;",
+        "org.apache.commons.codec.binary;Hex;true;Hex;(Charset);;Argument[0];Argument[-1];taint;",
+        "org.apache.commons.codec.binary;Hex;true;getCharset;();;Argument[-1];ReturnValue;taint;",
```

Comment:
I guess these work by themselves, but they aren't too relevant (a `Charset` won't carry taint most of the time).

I think some of the actually interesting models of the [Hex class](https://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/binary/Hex.html) are missing, though (basically, all the `encode` and `decode` methods).

Reply:
See super classes for the `encode`/`decode` methods

Reply:
Regarding `Chatset`. Good question. We'd somehow need to track certain classes as immutable and only instantiated under safe assumptions (e.g. no user input). Given `Charset` can be subclasses and could carry taint, I don't know if excluding this is the right way